test/e2e/upgrade_management_cluster_test.go

```diff
@@ -0,0 +1,336 @@
+		},
+	}
+	_, err = upgradeClusterClientSet.CoreV1().Namespaces().Create(ctx, ironicNamespace, metav1.CreateOptions{})
+	Expect(err).To(BeNil(), "Unable to create the Ironic namespace")
+
+	By("Configure Ironic Configmap")
+	configureIronicConfigmap(true)
+
+	By("Add labels to BMO CRDs")
+	labelBMOCRDs(nil)
+
+	By("Install BMO")
+	installIronicBMO(upgradeClusterProxy, "false", "true")
+
+	By("Install Ironic in the target cluster")
+	installIronicBMO(upgradeClusterProxy, "true", "false")
+
+	By("Restore original BMO configmap")
+	restoreBMOConfigmap()
+	By("Reinstate Ironic Configmap")
+	configureIronicConfigmap(false)
+
+	By("Add labels to BMO CRDs in the target cluster")
+	labelBMOCRDs(upgradeClusterProxy)
+
+	By("Ensure API servers are stable before doing move")
+	Consistently(func() error {
+		kubeSystem := &corev1.Namespace{}
+		return bootstrapClusterProxy.GetClient().Get(ctx, client.ObjectKey{Name: "kube-system"}, kubeSystem)
+	}, "5s", "100ms").Should(BeNil(), "Failed to assert bootstrap API server stability")
+	Consistently(func() error {
+		kubeSystem := &corev1.Namespace{}
+		return upgradeClusterClient.Get(ctx, client.ObjectKey{Name: "kube-system"}, kubeSystem)
+	}, "5s", "100ms").Should(BeNil(), "Failed to assert target API server stability")
+	By("Moving the cluster to self hosted")
+
+	cmd := exec.Command(clusterctlBinaryPath, "move", "--to-kubeconfig", upgradeClusterProxy.GetKubeconfigPath(), "-n", namespace, "-v", "10")
+	output, err := cmd.CombinedOutput()
+	Logf("move: %v", string(output))
+	Expect(err).To(BeNil())
+
+	By("Check if BMH is in provisioned state")
+	Eventually(func() error {
+		bmhList := &bmo.BareMetalHostList{}
+		if err := upgradeClusterClient.List(ctx, bmhList, client.InNamespace(namespace)); err != nil {
+			return err
+		}
+		for _, bmh := range bmhList.Items {
+			if !bmh.WasProvisioned() {
+				return errors.New("BMHs cannot be provisioned")
+			}
+		}
+		return nil
+	}, e2eConfig.GetIntervals(specName, "wait-bmh-provisioned")...).Should(BeNil())
+	// TODO (Mohammed) after merging template that can be parameterized from the e2e the cluster should boot with 1 kcp and remove this scale in part
+	Logf("Scale down KCP to 1 in order to 2 VMs for the workload cluster")
+	cmd = exec.Command("kubectl", "scale", "kcp", clusterName, "-n", namespace, "--replicas", "1", "--kubeconfig", upgradeClusterProxy.GetKubeconfigPath())
+	output, err = cmd.CombinedOutput()
+	Logf("scale: %v", string(output))
+	Expect(err).To(BeNil())
+
+	Logf("Waiting for 2 BMHs to be in Available state")
+	Eventually(func(g Gomega) {
+		bmhs, err := getAllBmhs(ctx, upgradeClusterClient, namespace, specName)
+		g.Expect(err).To(BeNil())
+		g.Expect(filterBmhsByProvisioningState(bmhs, bmo.StateAvailable)).To(HaveLen(2))
+	}, e2eConfig.GetIntervals(specName, "wait-bmh-available")...).Should(Succeed())
+
+	/*-------------------------------*
+	| Create a test workload cluster |
+	*--------------------------------*/
```

Comment:
I still think this part should be removed. Before pivoting was added to this upgrade test I said we needed it, but now that we have pivoting I don't think there is a point to keep it. It makes the e2e upgrade take longer time and makes it different from the ansible-based upgrade test.

The point is that we already test the old version by creating the first workload cluster and pivot to it. Then we do the upgrade with all controllers running in the (first) target cluster so that is also quite well tested.

Reply:
So this part might be an extra but since we are keeping the test steps close to CAPI I prefer to keep it since we are still working on reusing that test from CAPI

Reply:
After discussion we decided to merge as is. :heavy_check_mark: